### ISSUE TITLE
Query nested interfaces

### ISSUE THREAD

**Issue by joemasilotti** [nalexn/ViewInspector#37]
Do you have any thoughts on the amount of work required to _query_ nested interfaces? More like how UI Testing works?

Meaning, not having to write out every single intermediate view to get to a deeply nested one. Right now a complex hierarchy takes a lot of boilerplate to verify deeply nested items. Take, for example, the following.

```swift
// ContentView.swift

import SwiftUI

struct ContentView: View {
    var body: some View {
        NavigationView {
            VStack {
                NestedView(text: "First text.")
                NestedView(text: "Second text.")
            }
        }
    }
}

struct NestedView: View {
    var text: String

    var body: some View {
        HStack {
            Spacer()
            Image(systemName: "plus.app.fill")
            Text(text)
        }
    }
}
```

```swift
// ContentViewTests.swift

import XCTest
import ViewInspector
@testable import Ruka

class RukaTests: XCTestCase {
    func testNestedInterfaces() throws {
        let subject = ContentView()
        let secondText = try subject.inspect()
            .navigationView()
            .vStack(0)
            .view(NestedView.self, 1)
            .hStack()
            .text(2)
            .string()
        XCTAssertEqual(secondText, "Second text.")
    }
}

extension ContentView: Inspectable {}
extension NestedView: Inspectable {}
```

Ideally, I could write something like the following, more UI Testing style.

```swift
let secondText = try subject.inspect().texts["Second text."]
XCTAssert(secondText.exists)
```

---

I'm assuming that if this is possible it would require quite a bit of dynamic/meta programming since everything is nicely typed. Do you have any thoughts or ideas on this?

**Comment by nalexn** [comment]
Hey @joemasilotti 

That's an interesting idea, I'd need to learn more about your [repo](https://github.com/joemasilotti/UI-Testing-Cheat-Sheet) to get a better idea of the functionality that should come with this dynamic query.

At a high level, I think this should be possible. There is a function `Inspector.print(...)` that traverses the SwiftUI view tree and prints its structure, even by using this function and searching for `"Second text."` inside the output we already can assume the text is there. Of course, this requires more formal implementation, but the idea looks feasible.

You are welcomed to share any recommendations / caveats / hints on what should be there and how to best approach this!

**Comment by joemasilotti** [comment]
The repo is a cheat sheet for how to use the built in UI Testing functionality of Xcode. It doesn't actually "do" anything, it's just a few practical examples.

The dynamic querying is built in to UI Testing and has been since it was released in Xcode 7. To sum it up, you can query a nested view hierarchy without having to worry about the entire structure. Think of how a CSS query selector works for the web; you don't need to do `html header .bg-blue #my-div` to reference the div, just `#mydiv`.

---

`Inspector.print()` looks promising! I will have to dive a little deeper into the guts of ViewInspector to offer any solid advice or recommendations.

In terms of the API, it would ideally mimic UI Testing, as shown in the original comment. For example, all three of these should find the same `Text`.

```swift
// current
try subject.inspect()
    .navigationView()
    .vStack(0)
    .view(NestedView.self, 1)
    .hStack()
    .text(2)

// removing a few intermediary views should still work
try subject.inspect()
    .navigationView()
    .texts["Second text"]        

// finding a specific element by it's `string()` value
try subject.inspect().texts(string: "Second text.")

// syntactic sugur for above
try subject.inspect().texts["Second text."]
```

**Comment by Alexander-techIOS** [comment]
Not sure if creating a new ticket was the best option but since my query is related to this regarding nested interfaces.

I have a custom SwiftUI view (not UIViewRepresentable) it is presented on a state. 

```
if self.viewModel.displayInformation {
        Information(viewModel: viewModel)
            .transition(.move(edge: .bottom))
}
```

Triggering the that `displayInformation` is easy with `ViewInspector` via `.button().tap()` but there is an issue where the viewModel is reinitialized. 

`Information` has viewModel as `@StateObject`, am I doing something wrong with the SwifUI architecture or am I missing something in the way I use `ViewInspector` to test? If I build the app it works as it should on device.

I have tried `.inspect(onReceive:)` (in one of your examples) and I have also tried `.inspect(after:)` they all reinitialize the ViewModel wiping out any changes from before. I have noticed a purple flag (Xcode) saying something like StateObject used before installed on a view, which results in initialization on each access. But in this case I don't get that.

**Comment by nalexn** [comment]
@Alexander-techIOS I bet the problem in your case is that `@StateObject` is not yet officially supported by the library. I just didn't have time to get to it. This ticket is rather about the dynamic view lookup.

**Comment by Alexander-techIOS** [comment]
@nalexn Ah I see I see, yes I remember your docs somewhere pointing that out. It is working on one view, the moment it passes to another view it changes. So I guess in my mind I flagged it as it's working! haha. 

I must say I absolutely love your framework. When I get more familiar with SwiftUI and how your framework works, I would be very interested in contributing and helping out.

**Comment by nalexn** [comment]
Released in v0.6.0
